File: services/entity-resolution-service/src/fusion_base/models/cluster.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{FromRow, types::Json as SqlJson};
use uuid::Uuid;

use super::golden_record::GoldenRecord;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityRecord {
    pub record_id: String,
    pub source: String,
    pub external_id: String,
    pub display_name: String,
    pub confidence: f32,
    pub attributes: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MatchEvidence {
    pub left_record_id: String,
    pub right_record_id: String,
    pub blocking_key: String,
    pub rule_score: f32,
    pub ml_score: f32,
    pub final_score: f32,
    pub comparators: Vec<String>,
    pub explanation: String,
    pub requires_review: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReviewQueueItem {
    pub id: Uuid,
    pub cluster_id: Uuid,
    pub status: String,
    pub severity: String,
    pub recommended_action: String,
    pub rationale: Vec<String>,
    pub assigned_to: Option<String>,
    pub reviewed_by: Option<String>,
    pub notes: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResolvedCluster {
    pub id: Uuid,
    pub job_id: Uuid,
    pub cluster_key: String,
    pub status: String,
    pub records: Vec<EntityRecord>,
    pub evidence: Vec<MatchEvidence>,
    pub confidence_score: f32,
    pub requires_review: bool,
    pub suggested_golden_record_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ClusterDetail {
    pub cluster: ResolvedCluster,
    pub review_item: Option<ReviewQueueItem>,
    pub golden_record: Option<GoldenRecord>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SubmitReviewRequest {
    pub decision: String,
    pub notes: Option<String>,
    pub reviewed_by: Option<String>,
}

#[derive(Debug, Clone, FromRow)]
pub struct ClusterRow {
    pub id: Uuid,
    pub job_id: Uuid,
    pub cluster_key: String,
    pub status: String,
    pub records: SqlJson<Vec<EntityRecord>>,
    pub evidence: SqlJson<Vec<MatchEvidence>>,
    pub confidence_score: f32,
    pub requires_review: bool,
    pub suggested_golden_record_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow)]
pub struct ReviewQueueRow {
    pub id: Uuid,
    pub cluster_id: Uuid,
    pub status: String,
    pub severity: String,
    pub recommended_action: String,
    pub rationale: SqlJson<Vec<String>>,
    pub assigned_to: Option<String>,
    pub reviewed_by: Option<String>,
    pub notes: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<ClusterRow> for ResolvedCluster {
    fn from(value: ClusterRow) -> Self {
        Self {
            id: value.id,
            job_id: value.job_id,
            cluster_key: value.cluster_key,
            status: value.status,
            records: value.records.0,
            evidence: value.evidence.0,
            confidence_score: value.confidence_score,
            requires_review: value.requires_review,
            suggested_golden_record_id: value.suggested_golden_record_id,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

impl From<ReviewQueueRow> for ReviewQueueItem {
    fn from(value: ReviewQueueRow) -> Self {
        Self {
            id: value.id,
            cluster_id: value.cluster_id,
            status: value.status,
            severity: value.severity,
            recommended_action: value.recommended_action,
            rationale: value.rationale.0,
            assigned_to: value.assigned_to,
            reviewed_by: value.reviewed_by,
            notes: value.notes,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

File: services/entity-resolution-service/internal/models/cluster.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// EntityRecord mirrors the Rust fusion_base::models::cluster::EntityRecord.
//
// Rust attributes is `serde_json::Value` but is always an object in
// practice (built via `json!({...})`); we map it to `map[string]any`
// so callers can index with `record.Attributes[field]`.
type EntityRecord struct {
	RecordID    string         `json:"record_id"`
	Source      string         `json:"source"`
	ExternalID  string         `json:"external_id"`
	DisplayName string         `json:"display_name"`
	Confidence  float32        `json:"confidence"`
	Attributes  map[string]any `json:"attributes"`
}

// MatchEvidence mirrors fusion_base::models::cluster::MatchEvidence.
type MatchEvidence struct {
	LeftRecordID   string   `json:"left_record_id"`
	RightRecordID  string   `json:"right_record_id"`
	BlockingKey    string   `json:"blocking_key"`
	RuleScore      float32  `json:"rule_score"`
	MLScore        float32  `json:"ml_score"`
	FinalScore     float32  `json:"final_score"`
	Comparators    []string `json:"comparators"`
	Explanation    string   `json:"explanation"`
	RequiresReview bool     `json:"requires_review"`
}

// ReviewQueueItem mirrors fusion_base::models::cluster::ReviewQueueItem.
type ReviewQueueItem struct {
	ID                uuid.UUID `json:"id"`
	ClusterID         uuid.UUID `json:"cluster_id"`
	Status            string    `json:"status"`
	Severity          string    `json:"severity"`
	RecommendedAction string    `json:"recommended_action"`
	Rationale         []string  `json:"rationale"`
	AssignedTo        *string   `json:"assigned_to"`
	ReviewedBy        *string   `json:"reviewed_by"`
	Notes             string    `json:"notes"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
}

// ResolvedCluster mirrors fusion_base::models::cluster::ResolvedCluster.
type ResolvedCluster struct {
	ID                      uuid.UUID       `json:"id"`
	JobID                   uuid.UUID       `json:"job_id"`
	ClusterKey              string          `json:"cluster_key"`
	Status                  string          `json:"status"`
	Records                 []EntityRecord  `json:"records"`
	Evidence                []MatchEvidence `json:"evidence"`
	ConfidenceScore         float32         `json:"confidence_score"`
	RequiresReview          bool            `json:"requires_review"`
	SuggestedGoldenRecordID *uuid.UUID      `json:"suggested_golden_record_id"`
	CreatedAt               time.Time       `json:"created_at"`
	UpdatedAt               time.Time       `json:"updated_at"`
}

// ClusterDetail mirrors fusion_base::models::cluster::ClusterDetail.
type ClusterDetail struct {
	Cluster      ResolvedCluster  `json:"cluster"`
	ReviewItem   *ReviewQueueItem `json:"review_item"`
	GoldenRecord *GoldenRecord    `json:"golden_record"`
}

// SubmitReviewRequest mirrors fusion_base::models::cluster::SubmitReviewRequest.
type SubmitReviewRequest struct {
	Decision   string  `json:"decision"`
	Notes      *string `json:"notes,omitempty"`
	ReviewedBy *string `json:"reviewed_by,omitempty"`
}
